File: machine learning models/CAPM Beta using Linear Regression/Python/capmregression.py
from __future__ import division # to avoid round to zero in Python 2.xx
# This changes the meaning of the division operator so that it will return a floating point number 
# if needed to give a (closer to) precise answer. The historical behaviour is for x / y to return 
# an int if both x and y are ints, which usually forces the answer to be rounded down

for name in dir():
    if not name.startswith('_'):
        del globals()[name]

import numpy as np
import os
import reg_capm_lib as lib
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import Axes3D

def optimizeCost(X,y,theta,step,maxrun):
    m = len(y)
    cost_range = np.zeros((maxrun,1))

    for iter in range(0,maxrun):
        # h = np.dot(np.transpose(X),np.transpose(theta))
        h = X.dot(theta)
        grad = 1/m*(h - y).T.dot(X)    # grad is 1 x d
        # grad = 1/m*np.dot(np.transpose(h - y),X)    # grad is 1 x d
        theta = theta - step * grad.T
        cost_range[iter] = 1/2/m*(h-y).T.dot(h-y) ## ** is element-wise power of 2
    
    # return multiple variables in a tuple
    return theta,cost_range
	
def computeCost(X,y,theta):
    m = len(y)
    cost = 0
    h = np.dot(X,theta) 
    cost = 1/2/m*(h-y).T.dot(h-y)
    return(cost)	
	
	
data = np.loadtxt(r"C:\xxx\CAPMuniverse.csv",delimiter=",",skiprows=1)
(m,n) = data.shape # number of observations x number of variables

# y is the return of an individual stock; X is the return of the market
# 0th column is Date, 0-based in Python (1 based in R and MATLAB)
y = np.matrix(data[:, 12]-data[:, 14]).reshape(m,1) # 12th is YHOO, 14th is the risk-free rate
X = data[:, 13]-data[:, 14] # 13th is the market return

# ================== Gradient Descent =======================
X = np.matrix(np.c_[np.ones((m,1)), X]) # now add a column of 1 to X so it becomes [x0,x1]
maxrun = 1000000 # maximum number of iterations
step = 0.1
theta = np.matrix(np.zeros((2, 1))) # parameters for x0 and x1 respectively

# result returned in a tuple
theta,cost_range= lib.optimizeCost(X,y,theta,step,maxrun) # matrix form function

pred = X.dot(theta) # predicted y or the hypothesis
# =============== Plot the data and results =================
# plot y against X;
plt.plot(X[:,1], y,"r.",markersize=1,label="Training data") # red dots
plt.ylabel("Individual Security")
plt.xlabel("S&P500")

# now plot the regression line
plt.plot(X[:,1],pred,"b.",markersize=1,label="Predicted regression line") # note: this is 0-based
plt.legend(loc="upper left")
plt.show()

# now plot the cost
plt.plot(cost_range,"b.",markersize=1,label="Cost") # note: this is 0-based
plt.show()

File: machine learning models/CAPM Beta using Linear Regression/R/capmregression.R
# cleaning
rm(list=ls())
path="C:/xxx"
setwd(path)

library("plot3D")
library("plotly")

# use source if the functions are in another file
optimizeCost <- function(X,y,theta,step,maxrun) {
  m <- dim(X)[1]
  cost_range <- rep(0,length=maxrun) 
  
  for (iter in 1:maxrun) {        
    h = X%*%theta
    grad = 1/m * t(h-y) %*% X # grad is 1 x d
    theta = theta - step * t(grad)#
    cost_range[iter] = 1/2/m*sum((h-y)**2)
  }
  # return multiple variables using a list
  return(list(theta=theta,cost_range=cost_range)) 
}
computeCost <- function(X,y,theta) {
  h <- X%*%theta
  cost <- 1/(2*length(y))*sum((h-y)**2)
  return(cost)
}

data <- read.csv('../../data/CAPMuniverse.csv')
m = dim(data)[1] # number of observations x number of variables
n = dim(data)[2] # number of observations x number of variables

# y is the return of an individual stock; X is the return of the market
y = data[, 13]-data[, 15] # 13th is YHOO, 15th is the risk-free rate
X = data[, 14]-data[, 15] # 14th is the market return

# ================== Gradient Descent =======================
X = cbind(rep(1,m), X) # now add a column of 1 to X so it becomes [x0,x1]
maxrun = 1e+6 # maximum number of iterations
step = 0.1
theta = rep(0,2) # parameters for x0 and x1 respectively
r <- optimizeCost(X,y,theta,step,maxrun) # matrix form function
pred = X%*%r$theta # predicted y or the hypothesis

# =============== Plot the data and results =================
# plot y against X;
# pch: which symbol (http://www.statmethods.net/advgraphs/parameters.html)
# cex: symbol scaled 
dev.new()
plot(X[,2],y,ylim=range(c(y,pred)),type="p",pch=20,cex=0.1,col="red",ann=FALSE) 
par(new = TRUE) # here to plot multiple series on the same graph, ylim should be the same
# now plot the regression line
plot(X[,2],pred,ylim=range(c(y,pred)),type="p",pch=20,cex=0.1,col="blue",ann=FALSE)
legend(min(X[,2]), max(y), c("Training data", "Predicted regression line"),cex=0.8,col=c("red","blue"),pch=20)
title(xlab="Market return",col.lab="black")
title(ylab="Security return",col.lab="black")
title(main="CAPM Regression",col.main="black")

# plot the cost vs the number of iterations
dev.new()
plot(r$cost_range,ylab="Cost",pch=20,cex=0.1,xlab="Number of interations",col="blue",ann=FALSE)
title(xlab="Iterations",col.lab="black")
title(ylab="Cost",col.lab="black")
title(main="Cost vs iterations",col.main="black")

th0 <- seq(r$theta[1]-10,r$theta[1]+10,length=100)
th1 <- seq(r$theta[2]-10,r$theta[2]+10,length=100)
grid <- mesh(th0,th1)

m <- length(th0)
n <- length(th1)
cost <- matrix(0,m,n)
for (i in (1:m)){
  for (j in (1:n)) {
    cost[i,j] <- computeCost(X,y,matrix(c(th0[i],th1[j])))
  }
}

surf3D(x=grid$x,y=grid$y,z=cost,colkey=FALSE,bty="b2",main="Cost")

